nat-lab/tests/test_core_api.py

```diff
@@ -0,0 +1,294 @@
+import pytest
+from contextlib import AsyncExitStack
+from dataclasses import dataclass
+from utils.connection_util import ConnectionTag, new_connection_by_tag
+from uuid import UUID
+
+
+@dataclass
+class MachineData:
+    public_key: str
+    hardware_identifier: str
+    os: str
+    os_version: str
+
+
+def verify_uuid(uuid_to_test, version=4):
+    try:
+        uuid_obj = UUID(uuid_to_test, version=version)
+    except ValueError:
+        assert False, "Not a valid UUID"
+    assert str(uuid_obj) == uuid_to_test, "Not a valid UUID"
+
+
+def create_machine_data(public_key, hardware_identifier, os, os_version):
+    return MachineData(
+        public_key=public_key,
+        hardware_identifier=hardware_identifier,
+        os=os,
+        os_version=os_version,
+    )
+
+
+async def clean_up_machines(connection, server_host):
+    get_machines_process = await connection.create_process([
+        "curl",
+        "-X",
+        "GET",
+        f"{server_host}/v1/meshnet/machines",
+        "-H",
+        "Content-Type: application/json",
+    ]).execute()
+
+    response_data = json.loads(get_machines_process.get_stdout())
+
+    for machine in response_data:
+        await connection.create_process([
+            "curl",
+            "-X",
+            "DELETE",
+            f"{server_host}/v1/meshnet/machines/{machine['identifier']}",
+            "-H",
+            "Content-Type: application/json",
+        ]).execute()
+
+
+@pytest.fixture(name="machine_data")
+def fixture_machine_data(request):
+    return request.param
+
+
+@pytest.fixture(name="server_host")
+def fixture_server_host():
+    return "http://10.0.80.86:8080"
+
+
+@pytest.fixture(name="registered_machines")
+async def fixture_register_machine(server_host, machine_data):
+    async with AsyncExitStack() as exit_stack:
+        connection = await exit_stack.enter_async_context(
+            new_connection_by_tag(ConnectionTag.DOCKER_CONE_CLIENT_1)
+        )
+
+        await clean_up_machines(connection, server_host)
+
+        registered_machines = []
+
+        for data in machine_data:
+            payload = json.dumps(data.__dict__)
+
+            http_process = await connection.create_process([
+                "curl",
+                "-X",
+                "POST",
+                f"{server_host}/v1/meshnet/machines",
+                "-H",
+                "Content-Type: application/json",
+                "-d",
+                payload,
+            ]).execute()
+
+            response_data = json.loads(http_process.get_stdout())
+            registered_machines.append(response_data)
+
+        return registered_machines
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "machine_data",
+    [[
+        create_machine_data(
+            "test-public-key-1", "88cc55d2c6f64ff680aaa7235374411d", "linux", "5.10"
+        ),
+        create_machine_data(
+            "test-public-key-2", "98dd66e2d7f74ff780bbb8236485522e", "windows", "5.11"
+        ),
```

Comment:
What's the format of a hash and how it was created? If it's required to be derived from the key then it would be better to use `func('test-public-key-1')` instead of a hash :)

Reply:
Well, we don't have any validation on keys. From API perspective string type is expected.